nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/queryrecord/FlowFileTable.java

```diff
@@ -223,12 +225,69 @@ private RelDataType getRelDataType(final DataType fieldType, final JavaTypeFacto
             case BIGINT:
                 return typeFactory.createJavaType(BigInteger.class);
             case CHOICE:
+                final ChoiceDataType choiceDataType = (ChoiceDataType) fieldType;
+                DataType widestDataType = choiceDataType.getPossibleSubTypes().get(0);
+                for (final DataType possibleType : choiceDataType.getPossibleSubTypes()) {
+                    if (possibleType == widestDataType) {
+                        continue;
+                    }
+                    if (possibleType.getFieldType().isWiderThan(widestDataType.getFieldType())) {
+                        widestDataType = possibleType;
+                        continue;
+                    }
+                    if (widestDataType.getFieldType().isWiderThan(possibleType.getFieldType())) {
+                        continue;
+                    }
+
+                    // Neither is wider than the other.
+                    widestDataType = null;
+                    break;
+                }
+
+                // If one of the CHOICE data types is the widest, use it.
+                if (widestDataType != null) {
+                    return getRelDataType(widestDataType, typeFactory);
+                }
+
+                // None of the data types is strictly the widest. Check if all data types are numeric.
+                // This would happen, for instance, if the data type is a choice between float and integer.
+                // If that is the case, we can use a String type for the table schema because all values will fit
+                // into a String. This will still allow for casting, etc. if the query requires it.
+                boolean allNumeric = true;
+                for (final DataType possibleType : choiceDataType.getPossibleSubTypes()) {
+                    if (!isNumeric(possibleType)) {
+                        allNumeric = false;
+                        break;
+                    }
+                }
+
+                if (allNumeric) {
```

Comment:
Maybe allow String types too?

Reply:
I don't think we want to combine String types here. The idea here is to have a "parent" type that can be used for all numbers. Unfortunately, if the class is `Number.class` calcite does not recognize this, so it translates it into a SQL "OTHER" type. That prevents us still from doing things like SUM() on that column. By using a String type here, we can use SUM and other functions because all numbers can be converted to Strings and those particular Strings can also be converted to numbers under the hood.

However, if we allowed STRINGs to be included, it means we could have a schema whose field is a CHOICE between STRING and INT. It doesn't really make sense to perform aggregate functions such as SUM() across those values.

Or, said another way, String is the internal Java representation that we need to use in order for Calcite to allow both INT and FLOAT to be okay in the same column. But if our schema says that a value should be a CHOICE of STRING or INT, we shouldn't allow things like SUM() over that because you can't sum together STRINGs.

Reply:
As an example, consider a csv like:
```
name, other
markap14, 48
pcgrenier, computer
```
In this case, the schema has a field with name 'name' and a type String. But 'other' field is a CHOICE[INT, STRING]. So what should best represent that field in terms of java objects? I'd say `Object.class`. The `String.class` being used here is honestly a bit of a hack because Calcite doesn't give us a better way to represent Number - or, at least, not to my knowledge :)

Reply:
While I do agree, but given the example if you have split the records to sort out all the "number" values, the schema and the data could disagree. So logically you could know the data is all numbers but the schema disagrees. So if we are hacking to a string anyway, I don't think it would hurt to just allow it to pass on to let Calcite try and process it. This is just allowing a "best case" situation where Calcite isn't passed something it has no idea what to do with. So I do see a benefit in allowing the string to pass through, since if it is not a number it will still fail at query time but will at least attempt it.

Reply:
@pcgrenier I do see your argument. But I think I disagree. Let's say that we have the following CSV then:
```
name, other
markap14, 48
pcgrenier, 19
```
And the data's schema indicates that "other" is a CHOICE between STRING or INT. Fundamentally, it comes down to one question: in this case, should we allow the user to run a query like `SELECT SUM(other) as total FROM FLOWFILE`.

Your argument is yes, because I happen to know that for this particular FlowFile it will succeed.

My argument, though, is that we should not allow it. Yes, it would succeed for this FlowFile, but what will happen is that it will fail for other FlowFiles. Other FlowFiles that have the same schema and other FlowFiles for which the data is exactly correct according to the schema. This makes things more confusing because some FlowFiles succeed and others fail. All the while, the schema clearly indicates that the data may be a STRING, and you shouldn't really be able to sum together STRING data. If you do sort out all the number values, it makes sense to use a schema that reflects that. Otherwise, if we allow summing together a CHOICE[int, string] we're not really honoring the schema, in my point of view.